internal/common/spiretrustsource.go

```diff
@@ -0,0 +1,192 @@
+		localBackupDir:     localBackupDir,
+		domainCertificates: make(map[string][]*x509.Certificate, 0),
+		updateChan:         make(chan struct{}, 0),
+		updateTimeout:      5 * time.Second,
+	}
+
+	re := regexp.MustCompile(`^spiffe://(\S+)$`)
+	for trustDomain, spireURL := range spireEndpointURLs {
+		localStoragePath := ""
+		if localBackupDir != "" {
+			matches := re.FindStringSubmatch(trustDomain)
+			if len(matches) < 2 {
+				return nil, fmt.Errorf("expected domain of form spiffe://<trust_domain> but got %s", trustDomain)
+			}
+			domain := matches[1]
+			if strings.Contains(domain, "/") {
+				return nil, fmt.Errorf("expected domain without slash but got %s", domain)
+			}
+			localStoragePath = localBackupDir + "/" + domain + ".pem"
+		}
+
+		client, err := workload.NewX509SVIDClient(
+			&workloadWatcher{
+				domain:           trustDomain,
+				source:           source,
+				localStoragePath: localStoragePath,
+			},
+			workload.WithAddr(spireURL),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to construct a new NewX509SVIDClient for %s: %v", spireURL, err)
+		}
+
+		source.spireEndpoints[trustDomain] = &SpireEndpoint{
+			domain:    trustDomain,
+			loadState: Pending,
+			spireURL:  spireURL,
+			client:    client,
+		}
+	}
+
+	for _, spireEndpoint := range source.spireEndpoints {
+		if err := spireEndpoint.client.Start(); err != nil {
+			return nil, fmt.Errorf("failed to start NewX509SVIDClient for domain %s: %v", spireEndpoint.domain, err)
+		}
+	}
+
+	return source, nil
+}
+
+// Stop stops all spire clients
+func (s *SpireTrustSource) Stop() error {
+	errs := make([]string, 0)
+	for _, spireEndpoint := range s.spireEndpoints {
+		if err := spireEndpoint.client.Stop(); err != nil {
+			errs = append(errs, fmt.Sprintf("domain %s: %v", spireEndpoint.domain, err))
+		}
+	}
+
+	if len(errs) != 0 {
+		return fmt.Errorf("failed to stop NewX509SVIDClients: (%s)", strings.Join(errs, "),("))
+	}
+
+	return nil
+}
+
+func (w *workloadWatcher) UpdateX509SVIDs(svids *workload.X509SVIDs) {
+	certs := svids.Default().TrustBundle
+	w.source.domainCertificates[w.domain] = certs
+	w.source.spireEndpoints[w.domain].loadState = Loaded
+
+	if w.localStoragePath != "" {
+		builder := strings.Builder{}
+		for _, cert := range certs {
+			block := &pem.Block{
+				Type:  "CERTIFICATE",
+				Bytes: cert.Raw,
+			}
+			builder.Write(pem.EncodeToMemory(block))
+		}
+		file, err := appFS.OpenFile(w.localStoragePath, os.O_WRONLY|os.O_CREATE, 0600)
+		if err != nil {
+			logrus.Warnf("could not open backup file for trust domain %s at %s: %v", w.domain, w.localStoragePath, err)
+		} else {
+			defer file.Close()
+			_, err = file.WriteString(builder.String())
+			if err != nil {
+				logrus.Warnf("could not write to backup file for trust domain %s at %s: %v", w.domain, w.localStoragePath, err)
+			}
+		}
+	}
+
+	select {
+	case w.source.updateChan <- struct{}{}:
+	case <-time.After(w.source.updateTimeout):
+	}
```

Comment:
Is it necessary to write the backup to disk _before_ making the certs available for use? I think it might be better to make them available for other threads, then write to disk. This makes the disk write a non-blocking call for cert usage.

Reply:
I need this blocking so that I can make sure the write completed before reading a file in my tests. I'm not too worried about extra blocking in tests. However, I only need it in tests, so I've added a testing flag. Let me know what you think.